### DISCUSSION TITLE
Internal error when adding addressbook

### DISCUSSION THREAD

**Discussion by tfboy** [blind-coder/rcmcarddav#472] [Q&A]
Having installed carddav and got the plugin appearing in Roundcube, I've now hit a new issue.
I have created an "app" in Nextcloud which contains my addressbook. Using the name, app password and the URL provided by Nextcloud, as soon as I click Save, I get the "Oops" error again.
In the roundcube logs, I see this:
```
[30-May-2024 13:47:13 UTC] PHP Fatal error:  Uncaught Error: Call to undefined method GuzzleHttp\Utils::chooseHandler() in /var/lib/roundcube/vendor/guzzlehttp/guzzle/src/functions.php:61
Stack trace:
#0 /var/lib/roundcube/plugins/carddav/vendor/guzzlehttp/guzzle/src/HandlerStack.php(42): GuzzleHttp\choose_handler()
#1 /var/lib/roundcube/plugins/carddav/vendor/mstilkerich/carddavclient/src/HttpClientAdapterGuzzle.php(120): GuzzleHttp\HandlerStack::create()
#2 /var/lib/roundcube/plugins/carddav/vendor/mstilkerich/carddavclient/src/CardDavClient.php(58): MStilkerich\CardDavClient\HttpClientAdapterGuzzle->__construct()
#3 /var/lib/roundcube/plugins/carddav/vendor/mstilkerich/carddavclient/src/Account.php(143): MStilkerich\CardDavClient\CardDavClient->__construct()
#4 /var/lib/roundcube/plugins/carddav/vendor/mstilkerich/carddavclient/src/Account.php(216): MStilkerich\CardDavClient\Account->getClient()
#5 /var/lib/roundcube/plugins/carddav/vendor/mstilkerich/carddavclient/src/Services/Discovery.php(120): MStilkerich\CardDavClient\Account->findCurrentUserPrincipal()
#6 /var/lib/roundcube/plugins/carddav/carddav.php(1371): MStilkerich\CardDavClient\Services\Discovery->discoverAddressbooks()
#7 /var/lib/roundcube/plugins/carddav/carddav.php(595): carddav->determineAddressbooksToAdd()
#8 /usr/share/roundcube/program/lib/Roundcube/rcube_plugin_api.php(520): carddav->savePreferences()
#9 /usr/share/roundcube/program/actions/settings/prefs_save.php(164): rcube_plugin_api->exec_hook()
#10 /usr/share/roundcube/program/include/rcmail.php(282): rcmail_action_settings_prefs_save->run()
#11 /usr/share/roundcube/index.php(278): rcmail->action_handler()
#12 /var/lib/roundcube/public_html/index.php(26): include('...')
#13 {main}
  thrown in /var/lib/roundcube/vendor/guzzlehttp/guzzle/src/functions.php on line 61
```
The URL I use is `https://<Nextcloud FQDN>/remote.php/dav/addressbooks/users/<username>/contacts/`.
Guzzlehttp is installed and I can't see anything obviously wrong.
Any ideas?
Thank you!

**Comment by mstilkerich** [comment]
I guess you installed rcmcarddav from tarball and now have two incompatible versions of guzzle installed, one coming with roundcube and one coming with rcmcarddav. Install via composer to avoid such issues.

**Reply by tfboy** [reply]
I installed rcmcarddav via composer. I'm not sure if guzzle was there or not before, but it's possible it was installed as part of the roundcube which was installed seperately.
But if there's no particular configuration and you're suspecting a version compatibility, I'll look into that. Thanks.

edited to add, here are the versions I have installed:
- php (system) is 8.2.19
- php-guzzlehttp-guzzle/stable,now 7.4.5-1 all [installed,automatic]

**Reply by mstilkerich** [reply]
Please check if there is a vendor subdirectory in the plugins/carddav directory? There should only be a single vendor subdirectory in the roundcube root directory. If you install with composer, it is important to run composer in the roundcube root directory.
